### ISSUE TITLE
Cannot assign to FunctionType

### ISSUE THREAD

**Issue by novasenco** [microsoft/pyright#1952]
Is this desired behavior? In python, you can assign attributes to functions, since they are objects.

```python
from types import FunctionType

def dec(obj):
    if isinstance(obj, FunctionType):
        obj.foo = True # Cannot assign member "foo" for type "FunctionType" Member "foo" is unknown
        return obj
    else:
        def wrapper(func):
            func.foo = bool(obj)
            return func
        return wrapper
```

**Comment by erictraut** [comment]
Yes, this is intended behavior. You can arguably assign any attribute to any object in Python, but it's the job of a type checker to verify that the attribute is known and the assigned value doesn't violate the type of that attribute. For example:

```python
class A:
    pass

A().foo = 4 # Error: Cannot assign member "foo" for type "A"
```

This will run without generating an exception in Python, but `foo` is not a defined attribute of class `A`, so a type checker should emit an error in this case. The same is true of function objects.

Of course, you can always use a `cast` operator if you want to override the type checker.

```python
cast(Any, obj).foo = True
```

I'm not certain I understand what you're trying to do with the above code. If you care to provide more details, I might be able to suggest some other options.

**Comment by novasenco** [comment]
@erictraut 

I was merely making a decorator that attaches attributes to a function.

Using the example above:

```python3
from types import FunctionType

def dec(obj):
    if isinstance(obj, FunctionType):
        obj.foo = True # Cannot assign member "foo" for type "FunctionType" Member "foo" is unknown
        return obj
    else:
        def wrapper(func):
            func.foo = bool(obj)
            return func
        return wrapper

@dec
def bar():pass
print(bar.foo) #> True

@dec(False)
def baz():pass
print(baz.foo) #> False
```
